GoogleChromeLabs/wp-sitemaps
# Enhance the sitemap stylesheet to dynamically generate columns in the table without knowing a priori what elements will appear in the sitemap.

### Issue Number
Fixes #152

### Description
Well, I went ahead and took some time to work on this even tho no one has replied to the issue (I'm kind of compulsive that way :-).

This PR modifies the XSLT stylesheet for sitemaps to allow it to render columns in the HTML table output for all distinct children of `Q{http://www.sitemaps.org/schemas/sitemap/0.9}url` in the sitemap (see below for the meaning of the `Q{...}xyz` notation), without knowing a priori what children there may be.

The columns are ordered as follows:

1. elements defined in the sitemaps XML Schema come first, in the order they are specified in the content model for `Q{http://www.sitemaps.org/schemas/sitemap/0.9}url`, i.e., `URL`, `Last Modified`, `Change Frequency` and `Priority`.
2. After that will come "extension" elements in the `http://www.sitemaps.org/schemas/sitemap/0.9` namespace, ordered lexically by their `local-name()` (see #151 for why that is necessary, even though such sitemaps are invalid according to the XML Schema)
3. After that, come columns for all elements in extension namespaces (even though it is not **yet** possible to generate sitemap instances with elements in extension namespaces, again, see #151).  Columns for extension elements are ordered lexically by their namespace URI and then by their `local-name()`.

The new stylesheet is smart enough to deal with cases where different `Q{http://www.sitemaps.org/schemas/sitemap/0.9}url` elements have different children, e.g.

```
<urlset xmlns='http://www.sitemaps.org/schemas/sitemap/0.9'>
	<url>
		<loc>https://example.com/</loc>
	</url>
	<url>
		<loc>https://example.com/some-page/</loc>
		<lastmod>2020-03-01T18:27:27-07:00</lastmod>
	</url>
	<url>
		<loc>https://example.com/another-page/</loc>
		<changefreq>weekly</changefreq>
	</url>
	<url>
		<loc>https://example.com/yet-another-page/</loc>
		<foo xmlns='urn:sparrowhawkcomputing.com'>some value</foo>
	</url>
</urlset>
```

will produce a table like:

```
<table>
	<thead>
		<tr>
			<th>URL</th>
			<th>Last Modified</th>
			<th>Change Frequency</th>
			<th class='extension urn:sparrowhawkcomputing.com'>Foo</th>
		</tr>
	</thead>
	<tbody>
		<tr>
			<td><a href='https://example.com/'>https://example.com/</td>
			<td></td>
			<td></td>
			<td class='extension urn:sparrowhawkcomputing.com'></td>
		</tr>
		<tr>
			<td><a href='https://example.com/some-page/'>https://example.com/some-page/</td>
			<td>2020-03-01T18:27:27-07:00</td>
			<td></td>
			<td class='extension urn:sparrowhawkcomputing.com'></td>
		</tr>
		<tr>
			<td><a href='https://example.com/another-page/'>https://example.com/another-page/</td>
			<td></td>
			<td>weekly</td>
			<td class='extension urn:sparrowhawkcomputing.com'></td>
		</tr>
		<tr>
			<td><a href='https://example.com/yet-another-page/'>https://example.com/yet-another-page/</td>
			<td></td>
			<td></td>
			<td class='extension urn:sparrowhawkcomputing.com'>some value</td>
		</tr>
	</tbody>
</table>
```

That example also illustrates that a `class` attribute is added for columns resulting from extension elements in case the plugins that add them want to style such columns differently (to make it clear to users that they extensions).  Of course, they won't be able to use the `class` shorthand notation in the CSS selectors, they'll have to do `th[class~="urn:sparrowhawkcomputing.com"], td[class~="urn:sparrowhawkcomputing.com"] { color: red; }` since the namespace URI's will contain characters not legal in the `class` shorthand selector syntax.

Because no browsers currently support anything other than XSLT/XPath 1.0, the stylesheet is a little more complicated than it would be if they supported XSLT/XPath 2.0/3.0.  In particular, it requires using one of 2 extension functions, [exsl:node-set()](http://exslt.org/exsl/functions/node-set/) or [msxsl:node-set()](https://docs.microsoft.com/en-us/dotnet/standard/data/xml/support-for-the-msxsl-node-set-function).  The first is supported by Chrome, Firefox, Safari, Opera and other browsers; the later is supported by Edge and IE (11).  As a fallback in case neither function is available, then the stylesheet will just render a single `URL` column.

We could (should?) get even more elaborate by providing a hook so that plugins that add extension elements to a sitemap can provide translatable text to use in the column heading for their extension elements (instead of just upper-casing the first letter of the `local-name()` as done here...but we can always add that later.

Here's a few miscelaneous notes:

* I've included inline comments in the new stylesheet to explain what it's doing...hopefully they'll make sense.  There is an equivalent of PHPDoc for XSLT called [XSLTDoc](https://github.com/XSLTdoc/XSLTdoc).  I didn't bother to use it, for now, but it something to consider in the long run.
* the `Q{...}xyz` notation used here (and in comments in the stylesheet) is a way to refer to namespace qualified elements independent of the prefix used for that namespace in any particular XML instance.  See the [URIQualifiedName](https://www.w3.org/TR/xpath-30/#doc-xpath30-URIQualifiedName) production in the XPath 3.0 spec.

### Type of change
Please select the relevant options:


### Steps to test
I've tested this in Chrome, Firefox, Edge and IE (11) on Windows.  Would really appreciate folks testing it on Mac & Linux with as many different browsers as you can!!

To test, you can do something like:

```
add_filter( 'core_sitemaps_posts_url_list', function( $url_list ) {
	$changefreq         = array( 'daily', 'weekly', 'monthly', 'yearly' );
	
	// this will result in the Change Frequency and Priority columns
	// appearing (or not) randomly each time you load the sitemap
	// to test that only columns for elements that appear in the sitemap
	// appear in the resulting output.
	$include_changefreq = random_int( 0, 1 );
	$include_priority   = random_int( 0, 1 );

	foreach ( $url_list as $idx => &$url ) {
		if ( $include_changefreq ) {
			$url['changefreq'] = $changefreq[ random_int( 0, count( $changefreq ) - 1 ) ];
		}
		if ( $include_priority ) {
			$url['priority'] = random_int( 0, 10 ) / 10;
		}
		// the random_int() calls below are to simulate a plugin
		// that only adds a key when a post has a specific post meta key (or meta value)
		// and will test that appropriate empty td elements are output so that
		// values end up in the correct columns.
		if ( random_int( 0, 1 ) ) {
			$url['foo'] = "foo:{$idx}";
		}
		if ( random_int( 0, 1 ) ) {
			$url['bar'] = "bar:{$idx}";
		}
	}

	return $url_list;
} );
```
### Acceptance criteria
